training/tf/tfprocess.py

```diff
@@ -544,28 +537,58 @@ def batch_norm(self, net):
             net = tf.layers.batch_normalization(
                     net,
                     epsilon=1e-5, axis=1, fused=True,
-                    center=True, scale=False,
+                    center=True, scale=scale,
                     training=self.training,
                     reuse=self.reuse_var)
 
-        for v in ['beta', 'moving_mean', 'moving_variance' ]:
+        for v in ['gamma', 'beta', 'moving_mean', 'moving_variance' ]:
+            if v == 'gamma' and not scale:
+                continue
             name = "fp32_storage/" + scope + '/batch_normalization/' + v + ':0'
             var = tf.get_default_graph().get_tensor_by_name(name)
             self.add_weights(var)
 
         return net
 
-    def conv_block(self, inputs, filter_size, input_channels, output_channels, name):
-        W_conv = weight_variable(
-            name,
-            [filter_size, filter_size, input_channels, output_channels],
-            self.model_dtype)
+    def squeeze_excitation(self, x, channels, ratio, name):
+
+        assert channels % ratio == 0
+
+        # NCHW format reduced to NC
+        net = tf.reduce_mean(x, axis=[2, 3])
+
+        W_fc1 = weight_variable(name + '_se_fc1_w', [channels, channels // ratio], self.model_dtype)
+        b_fc1 = bias_variable(name + '_se_fc1_b', [channels // ratio], self.model_dtype)
+        self.weights.append(W_fc1)
+        self.weights.append(b_fc1)
+
+        net = tf.nn.relu(tf.add(tf.matmul(net, W_fc1), b_fc1))
+
+        W_fc2 = weight_variable(name + '_se_fc2_w',
+            [channels // ratio, 2 * channels], self.model_dtype)
+        b_fc2 = bias_variable(name + '_se_fc2_b', [2 * channels], self.model_dtype)
+        self.weights.append(W_fc2)
+        self.weights.append(b_fc2)
 
-        self.add_weights(W_conv)
+        net = tf.add(tf.matmul(net, W_fc2), b_fc2)
+        net = tf.reshape(net, [-1, 2 * channels, 1, 1])
+
+        # Split to scale and bias
+        gammas, betas = tf.split(net, 2, axis=1)
+
+        out = tf.nn.sigmoid(gammas) * x + betas
```

Comment:
These are SE-FiLM units rather than the ones from the original paper, correct?

Reply:
Yes, they have a bias term calculated from pooling. This form was better in lc0 and setting biases to zero gives the original SE form. Performance difference is essentially zero.